clang/lib/AST/ASTContext.cpp

```diff
@@ -831,6 +831,28 @@ ASTContext::getCanonicalTemplateTemplateParmDecl(
   return CanonTTP;
 }
 
+/// Check if a type can have its sanitizer instrumentation elided.
+/// Determine this by its presence in a SCL alongside its specified categories.
+/// For example:
+/// ignorelist.txt>
+/// [{unsigned-integer-overflow,signed-integer-overflow}]
+/// type:*=no_sanitize
+/// type:size_t=sanitize
+/// <ignorelist.txt
+/// Supplying the above ignorelist.txt will disable overflow sanitizer
+/// instrumentation for all types except "size_t".
+bool ASTContext::isTypeIgnoredBySanitizer(const SanitizerMask &Mask,
+                                          const QualType &Ty) const {
+  bool sanitizeType =
+      NoSanitizeL->containsType(Mask, Ty.getAsString(), "sanitize");
+
+  bool noSanitizeType =
```

Comment:
Probably it should use `Ty.getAsString(getPrintingPolicy())`

Reply:
maybe even `Ty.getUnqualifiedType().getAsString(getPrintingPolicy())`

Reply:
Ok great. I'll use this as long as no regressions.

Out of curiosity: What is a printing policy?

Reply:
We do not want the canonical type, we should support typedefs.

How about this?

```c
Ty.getUnqualifiedType().getAsString(getPrintingPolicy());
```

Reply:
> Out of curiosity: What is a printing policy?

Not sure, but other users of containsType use policy.

> We do not want the canonical type, we should support typedefs.

Can you add a test to control that? with current patch canonical works.

Actually I was inspired by the patch to add -fsanitize=enum type ignore list, and 
without canonical `void f(enum MyEnum e)` is matched as `enum MyEnum`.

Reply:
Ok since some users of `isTypeIgnoredBySanitizer()` may want to use different type spellings for their purposes, should that method accept a `StringRef` which can be conjured up by any means necessary from the callsite instead of within the method itself?

@vitalybuka - pinging you since your notifications are strict :)

Reply:
I tried insert "getCanonicalType()" and none of tests of this patch failed.

Reply:
I guess qual typ, as-is,  is OK, and we can left out `getCanonicalType()` and let callers to insert that.
So `Ty.getUnqualifiedType().getAsString(getPrintingPolicy());` is LGTM

Reply:
@vitalybuka 
> I tried insert "getCanonicalType()" and none of tests of this patch failed.

TEST 'Clang :: CodeGen/ubsan-type-ignorelist-category.test' FAILED with this diff:

```diff
diff --git a/clang/lib/AST/ASTContext.cpp b/clang/lib/AST/ASTContext.cpp
index 480233894995..d66533d8a3d7 100644
--- a/clang/lib/AST/ASTContext.cpp
+++ b/clang/lib/AST/ASTContext.cpp
@@ -843,7 +843,8 @@ ASTContext::getCanonicalTemplateTemplateParmDecl(
 /// instrumentation for all types except "size_t".
 bool ASTContext::isTypeIgnoredBySanitizer(const SanitizerMask &Mask,
                                           const QualType &Ty) const {
-  std::string TyName = Ty.getUnqualifiedType().getAsString(getPrintingPolicy());
+  std::string TyName = Ty.getCanonicalType().getUnqualifiedType().getAsString(
+      getPrintingPolicy());
 
   return (NoSanitizeL->containsType(Mask, TyName) ||
           NoSanitizeL->containsType(Mask, TyName, "no_sanitize")) &&

```

log: https://gist.github.com/JustinStitt/bb3151fe54290966babcc1a9e78c8769

Reply:
I think we can refine in future. I don't have strong opinion on `getCanonicalType()/getUnqualifiedType()`